File: src/TlkFile.java
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class TlkFile  {

    TlkHeader header;
    public List<TlkStringRef> stringRefs;
    List<HuffmanNode> characterTree;
    BitArray bits;

    private enum FileFormat {
        txt,
        csv,
        xml
    }

    public class TlkHeader {
        public int magic;
        public int ver;
        public int min_ver;
        public int entry1Count;
        public int entry2Count;
        public int treeNodeCount;
        public int dataLen;

        public TlkHeader(FileInputStream r) throws IOException {
            this.magic = r.read();
            this.ver = r.read();
            this.min_ver = r.read();
            this.entry1Count = r.read();
            this.entry2Count = r.read();
            this.treeNodeCount = r.read();
            this.dataLen = r.read();
        }
    }

        public class TlkStringRef {
            public int stringId;
            public int BitOffset;

            public String Data;
            public int position;

            public TlkStringRef(FileInputStream r) throws IOException {
                this.stringId = r.read();
                this.BitOffset = r.read();
            }
        }

        public class HuffmanNode {
            public int leftNodeId;
            public int rightNodeId;

            public HuffmanNode(FileInputStream r) throws IOException {
                this.leftNodeId = r.read();
                this.rightNodeId = r.read();
            }
        }

        public void ProgressChangedEventHandler(int percentProgress);
        public Event progressChanged;

        private void OnProgressChanged(int percentProgress)
        {
            ProgressChangedEventHandler handler = ProgressChanged;
            if (handler != null)
                handler(percentProgress);
        }

        /// <summary>
        /// Loads a TLK file into memory.
        /// </summary>
        /// <param name="fileName"></param>
        public void LoadTlkData(String fileName, boolean isPC) throws IOException {
            /* **************** STEP ONE ****************
             *          -- load TLK file header --
             *
             * reading first 28 (4 * 7) bytes
             */

            /* using LittleEndian for PC architecture and BigEndian for Xbox360 */
            MiscUtil.Conversion.EndianBitConverter bitConverter;
            if (isPC)
                bitConverter = new MiscUtil.Conversion.LittleEndianBitConverter();
            else
                bitConverter = new MiscUtil.Conversion.BigEndianBitConverter();
            FileInputStream r = new MiscUtil.IO.EndianBinaryReader(bitConverter, File.OpenRead(fileName));

            header = new TlkHeader(r);
            /* read possibly correct ME2 TLK file, but from another platfrom */
            if (header.magic == 1416391424)
                throw new RuntimeException();
            /* read definately NOT a ME2 TLK ile */
            if (header.magic != 7040084)
                throw new RuntimeException();

            //DebugTools.PrintHeader(Header);

            /* **************** STEP TWO ****************
             *  -- read and store Huffman Tree nodes --
             */
            /* jumping to the beginning of Huffmann Tree stored in TLK file */
            long pos = r.BaseStream.Position;
            r.BaseStream.Seek(pos + (Header.entry1Count + Header.entry2Count) * 8, SeekOrigin.Begin);

            characterTree = new LinkedList<>();
            for (int i = 0; i < header.treeNodeCount; i++)
                characterTree.add(new HuffmanNode(r));

            /* **************** STEP THREE ****************
             *  -- read all of coded data into memory --
             */
            byte[] data = new byte[Header.dataLen];
            r.BaseStream.Read(data, 0, data.Length);
            /* and store it as raw bits for further processing */
            Bits = new BitArray(data);

            /* rewind BinaryReader just after the Header
             * at the beginning of TLK Entries data */
            r.BaseStream.Seek(pos, SeekOrigin.Begin);

            /* **************** STEP FOUR ****************
             * -- decode (basing on Huffman Tree) raw bits data into actual strings --
             * and store them in a Dictionary<int, string> where:
             *   int: bit offset of the beginning of data (offset starting at 0 and counted for Bits array)
             *        so offset == 0 means the first bit in Bits array
             *   string: actual decoded String */
            Map<Integer, String> rawStrings = new HashMap<>();
            int offset = 0;
            while (offset < Bits.Length)
            {
                int key = offset;
                /* read the String and update 'offset' variable to store NEXT String offset */
                String s = GetString(ref offset);
                rawStrings.Add(key, s);
            }

            /* **************** STEP FIVE ****************
             *         -- bind data to String IDs --
             * go through Entries in TLK file and read it's String ID and offset
             * then check if offset is a key in rawStrings and if it is, then bind data.
             * Sometimes there's no such key, in that case, our String ID is probably a substring
             * of another String present in rawStrings.
             */
            stringRefs = new LinkedList<>();
            for (int i = 0; i < header.entry1Count + header.entry2Count; i++) {
                TlkStringRef sref = new TlkStringRef(r);
                sref.position = i;
                if (sref.BitOffset >= 0)
                {
                    if (!rawStrings.containsKey(sref.BitOffset))
                    {
                        int tmpOffset = sref.BitOffset;
                        String partString = GetString(ref tmpOffset);

                        /* actually, it should store the fullString and subStringOffset,
                         * but as we don't have to use this compression feature,
                         * we will store only the part of String we need */

                        /* int key = rawStrings.Keys.Last(c => c < sref.BitOffset);
                         * String fullString = rawStrings[key];
                         * int subStringOffset = fullString.LastIndexOf(partString);
                         * sref.StartOfString = subStringOffset;
                         * sref.Data = fullString;
                         */
                        sref.Data = partString;
                    }
                    else
                    {
                        sref.Data = rawStrings[sref.BitOffset];
                    }
                }
                stringRefs.add(sref);
            }
            r.close();
        }

        /// <summary>
        /// Writes data stored in memory to an appriopriate text format.
        /// </summary>
        /// <param name="fileName"></param>
        /// <param name="ff"></param>
        public void DumpToFile(String fileName, FileFormat ff)
        {
            File.Delete(fileName);
            /* for now, it's better not to sort, to preserve original order */
            // StringRefs.Sort(CompareTlkStringRef);

            if (ff.equals(FileFormat.xml))
                SaveToXmlFile(fileName);
            else
                SaveToTextFile(fileName);
        }

        /// <summary>
        /// Starts reading 'Bits' array at position 'bitOffset'. Read data is
        /// used on a Huffman Tree to decode read bits into real strings.
        /// 'bitOffset' variable is updated with last read bit PLUS ONE (first unread bit).
        /// </summary>
        /// <param name="bitOffset"></param>
        /// <returns>
        /// decoded String or null if there's an error (last string's bit code is incomplete)
        /// </returns>
        /// <remarks>
        /// Global variables used:
        /// List(of HuffmanNodes) CharacterTree
        /// BitArray Bits
        /// </remarks>
        private String GetString(ref int bitOffset)
        {
            HuffmanNode root = characterTree.get(0);
            HuffmanNode curNode = root;

            String curString = "";
            int i;
            for (i = bitOffset; i < Bits.Length; i++)
            {
                /* reading bits' sequence and decoding it to Strings while traversing Huffman Tree */
                int nextNodeID;
                if (Bits[i])
                    nextNodeID = curNode.RightNodeID;
                else
                    nextNodeID = curNode.LeftNodeID;

                /* it's an internal node - keep looking for a leaf */
                if (nextNodeID >= 0)
                    curNode = CharacterTree[nextNodeID];
                else
                    /* it's a leaf! */
                {
                    char c = BitConverter.ToChar(BitConverter.GetBytes(0xffff - nextNodeID), 0);
                    if (c != '\0')
                    {
                        /* it's not NULL */
                        curString += c;
                        curNode = root;
                    }
                    else
                    {
                        /* it's a NULL terminating processed string, we're done */
                        bitOffset = i + 1;
                        return curString;
                    }
                }
            }
            bitOffset = i + 1;
            return null;
        }

        /// <summary>
        /// Writing data in an XML format.
        /// </summary>
        /// <param name="fileName"></param>
        private void SaveToXmlFile(String fileName)
        {
            int totalCount = stringRefs.size();
            int count = 0;
            int lastProgress = -1;
            XmlTextWriter xr = new XmlTextWriter(fileName, Encoding.UTF8);
            xr.Formatting = Formatting.Indented;
            xr.Indentation = 4;

            xr.WriteStartDocument();
            xr.WriteStartElement("tlkFile");
            xr.WriteAttributeString("TLKToolVersion", App.GetVersion());

            xr.WriteComment("Male entries section begin (ends at position " + (Header.entry1Count - 1) + ")");

            for (TlkStringRef s : stringRefs) {
                if (s.position == Header.entry1Count)
                {
                    xr.WriteComment("Male entries section end");
                    xr.WriteComment("Female entries section begin (ends at position " + (Header.entry1Count + Header.entry2Count - 1) + ")");
                }

                xr.WriteStartElement("string");

                xr.WriteStartElement("id");
                xr.WriteValue(s.StringID);
                xr.WriteEndElement(); // </id>

                xr.WriteStartElement("position");
                xr.WriteValue(s.position);
                xr.WriteEndElement(); // </position>

                if (s.BitOffset < 0)
                    xr.WriteElementString("data", "-1");
                else
                    xr.WriteElementString("data", s.Data);

                xr.WriteEndElement(); // </string>

                int progress = (++count * 100) / totalCount;
                if (progress > lastProgress)
                {
                    lastProgress = progress;
                    OnProgressChanged(lastProgress);
                }
            }
            xr.WriteComment("Female entries section end");
            xr.WriteEndElement(); // </tlkFile>
            xr.Flush();
            xr.Close();
        }

        /** <summary>
         * Writing data in a normal text format.
         * </summary>
         * <remarks>
         * Currently not used by main application, but it works ok.
         * </remarks>
         * <param name="fileName"></param>
         */
        private void SaveToTextFile(String fileName)
        {
            int totalCount = stringRefs.size();
            int count = 0;
            int lastProgress = -1;

            for (TlkStringRef s : stringRefs) {
                String line = s.StringID + ": " + s.Data + "\r\n";
                File.AppendAllText(fileName, line);

                int progress = (++count * 100) / totalCount;
                if (progress > lastProgress)
                {
                    lastProgress = progress;
                    OnProgressChanged(lastProgress);
                }
            }
        }

//        /* for sorting */
//        private static int CompareTlkStringRef(TlkHeader.TlkStringRef strRef1, TlkHeader.TlkStringRef strRef2) {
//            int result = strRef1.stringId.compareTo(strRef2.stringId);
//            return result;
//        }
    }
}

File: python/tlk_file.py
import os

import bit_array
import bit_convertor
import huffman_node
import tlk_header
import input_stream
import tlk_string_ref
import wrap
from xml.etree import ElementTree as ET


class TlkFile:
    def __init__(self):
        self.header = None
        self.string_refs = []
        self.character_tree = []
        self.bits = []

    def get_string(self, bit_offset_wrap):
        root = self.character_tree[0]
        cur_node = root
        cur_string = ''

        i = bit_offset_wrap.value
        while i < self.bits.length:
            next_node_id = cur_node.right_node_id if self.bits.get_rev(i) else cur_node.left_node_id

            if next_node_id >= 0:
                cur_node = self.character_tree[next_node_id]
            else:
                c = chr(0)
                try:
                    c = bit_convertor.to_char_rev(bit_convertor.get_bytes_by_value(0xffff - next_node_id), 0)
                except:
                    raise Exception
                if c != '\0':
                    cur_string += c
                    cur_node = root
                else:
                    i += 1
                    bit_offset_wrap.value = i
                    return cur_string
            i +=1
        i += 1
        bit_offset_wrap.value = i
        return None

    def load_tlk_data(self, source_path):
        input_s = input_stream.InputStream(source_path)
        self.header = tlk_header.TlkHeader(input_s)

        if self.header.magic == 1416391424:
            raise Exception('header.magic == 1416391424')
        if self.header.magic != 7040084:
            raise Exception('header.magic != 7040084')

        pos = input_s.pos
        r = input_stream.InputStream(pos, source_path)
        r.pos = pos + (self.header.entry_1_count + self.header.entry_2_count) * 8

        for i in range(self.header.tree_node_count):
            h_node = huffman_node.HuffmanNode(r)
            self.character_tree += [h_node]

        data_length = self.header.data_len
        data = [None] * data_length
        r.read_to_array(data, 0, data_length)
        self.bits = bit_array.BitArray(a=data)

        r = input_stream.InputStream(pos, source_path)

        raw_strings = {}
        offset = 0
        offset_wrap = wrap.Wrap(offset)

        while offset_wrap.value < self.bits.length:
            key = offset_wrap.value
            s = self.get_string(offset_wrap)
            raw_strings[key] = s

        for i in range(self.header.entry_1_count + self.header.entry_2_count):
            s_ref = tlk_string_ref.TlkStringRef(r)
            s_ref.position = i
            if s_ref.bit_offset >= 0:
                s_ref.data = raw_strings[s_ref.bit_offset] \
                    if s_ref.bit_offset in raw_strings.keys() \
                    else self.get_string(wrap.Wrap(s_ref.bit_offset))
            self.string_refs.append(s_ref)

    def store_to_file(self, dest_file, file_format):
        if os.path.isfile(dest_file):
            os.remove(dest_file)
        if file_format == 'xml':
            self.save_to_xml_file(dest_file)
        #             pretty_xml(dest_file)
        else:
            # save_to_text_file(dest_file)
            pass

    def save_to_xml_file(self, abs_path):
        # doc = ET.fromstring("<test>test öäü</test>")

        root = ET.Element('tlkFile')
        root.set("TLKToolVersion", '1.0.4')
        comment = ET.Comment(
            'Male entries section begin (ends at position {0})'.format((str(self.header.entry_1_count - 1))))
        root.append(comment)

        for i in range(len(self.string_refs)):
            sr = self.string_refs[i]

            if sr.position == self.header.entry_1_count:
                comment1 = ET.Comment('Male entries section end')
                comment2 = ET.Comment('Female entries section begin (ends at position {0})'.format((str(self.header.entry_1_count + self.header.entry_2_count - 1))))
                root.append(comment1)
                root.append(comment2)

            s = ET.SubElement(root, 'string')

            s1 = ET.SubElement(s, 'id')
            s1.text = str(sr.string_id)

            s2 = ET.SubElement(s, 'position')
            s2.text = str(sr.position)

            s3 = ET.SubElement(s, 'data')
            s3.text = '-1' if sr.bit_offset < 0 else sr.data
        comment = ET.Comment("Female entries section end")
        root.append(comment)

        tree = ET.ElementTree(root)
        ET.indent(tree, space="\t", level=0)
        tree.write(abs_path, encoding="utf-8")
